src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -518,20 +523,55 @@ public CheckConstraintsArgs(CSharpCompilation currentCompilation, ConversionsBas
             }
         }
 
+        private static readonly ObjectPool<CheckConstraintsArgsBoxed> s_checkConstraintsArgsBoxedPool = new ObjectPool<CheckConstraintsArgsBoxed>(() => new CheckConstraintsArgsBoxed());
+
         internal sealed class CheckConstraintsArgsBoxed
         {
-            public readonly CheckConstraintsArgs Args;
+            public CheckConstraintsArgs Args;
 
             [MethodImpl(MethodImplOptions.NoInlining)]
-            public CheckConstraintsArgsBoxed(CSharpCompilation currentCompilation, ConversionsBase conversions, Location location, BindingDiagnosticBag diagnostics)
+            public static PooledCheckConstraintsArgsBoxed Create(CSharpCompilation currentCompilation, ConversionsBase conversions, Location location, BindingDiagnosticBag diagnostics)
             {
-                Args = new CheckConstraintsArgs(currentCompilation, conversions, location, diagnostics);
+                var boxedArgs = s_checkConstraintsArgsBoxedPool.Allocate();
+                boxedArgs.Args = new CheckConstraintsArgs(currentCompilation, conversions, location, diagnostics);
+                return new PooledCheckConstraintsArgsBoxed(boxedArgs);
             }
 
             [MethodImpl(MethodImplOptions.NoInlining)]
-            public CheckConstraintsArgsBoxed(CSharpCompilation currentCompilation, ConversionsBase conversions, bool includeNullability, Location location, BindingDiagnosticBag diagnostics)
+            public static PooledCheckConstraintsArgsBoxed Create(CSharpCompilation currentCompilation, ConversionsBase conversions, bool includeNullability, Location location, BindingDiagnosticBag diagnostics)
+            {
+                var boxedArgs = s_checkConstraintsArgsBoxedPool.Allocate();
+                boxedArgs.Args = new CheckConstraintsArgs(currentCompilation, conversions, includeNullability, location, diagnostics);
+                return new PooledCheckConstraintsArgsBoxed(boxedArgs);
+            }
+
+            internal ref struct PooledCheckConstraintsArgsBoxed
```

Comment:
Do we need both types: `CheckConstraintsArgsBoxed` and `PooledCheckConstraintsArgsBoxed`? Could we simplify down to a single type?

Reply:
not without vioalting the compiler rules about 'using' pooled types :)  i'm happy to merge if that's ok (or i could move to manual free).

Reply:
merged.